src/BindPattern.jl

```diff
@@ -272,53 +283,28 @@ function bind_pattern!(
     return (pattern, assigned)
 end
 
-function push_pattern!(patterns::Vector{BoundPattern}, state::BinderState, pat::BoundFetchPattern)
+function push_pattern!(
+    patterns::Vector{BoundPattern},
+    state::BinderState,
+    pat::BoundFetchPattern
+)
     temp = get_temp(state, pat)
     push!(patterns, pat)
     temp
 end
 
 #
-# Infer which fields to match in a positional struct pattern by inspecting the set
-# of constructors.  It would be nice to exclude constructors that have
-# required keyword parameters, but the Julia APIs offer no simple way to determine
-# which keyword parameters have defaults.  That's because keyword parameters without
-# defaults are just rewritten into keyword parameters with defaults that throw an
-# exception at runtime.  So we exclude functions that have any keyword parameters.
-# If that ends up being problematic, we'll revisit the strategy.
+# Return a tuple containing the ordered list of the names (as Symbols) of fields that
+# can be matched either nominally or positionally.  This list should exclude synthetic
+# fields that are produced by packages such as Mutts and AutoHashEqualsCached.  This
+# function may be overridden by the client to hide fields that should not be matched.
 #
-function infer_fieldnames(type::Type, len::Int, match_positionally::Bool, location::LineNumberNode)
-    members = try
-        fieldnames(type)
-    catch ex
-        error("$(location.file):$(location.line): Could not determine the field names of `$type`.")
-    end
-
-    # If we're matching by keyword, we permit the use of any declared fields.
-    match_positionally || return members
-
-    # Search for constructor methods that have the correct number of parameters,
-    # no keyword parameters, and are not varargs.
-    meths = Method[methods(type)...]
-    meths = filter(m -> !m.isva && length(Base.kwarg_decl(m))==0, meths)
-    # drop the implicit var"#self#" argument
-    argnames = map(m -> dropfirst(Base.method_argnames(m)), meths)
-    # narrow to arg lists of the correct length where all parameter names correspond to members
-    argnames = unique(filter(l -> length(l) == len && all(n -> n in members, l), argnames))
-
-    if length(argnames) == 1
-        # found a uniquely satisfying order for member names
-        return (argnames[1]...,)
-    elseif len == length(members)
-        # no unique constructor, but the correct number of fields exist; use them
-        return members
-    elseif len > length(members)
-        error("$(location.file):$(location.line): The type `$type` has $(length(members)) fields but the pattern expects $len fields.")
-    else
-        error("$(location.file):$(location.line): Cannot infer which $len of the $(length(members)) fields to match from any positional constructor for `$type`.")
-    end
+function fieldnames(type::Type)
```

Comment:
maybe `matchable_fieldnames`? It's a bit longer/clunkier, but more explicit and unique

Reply:
I'd rather do that in a separate PR, since I already have code building on this.  Filed #56.